### PR TITLE
Bug/ux reloading pages

### PR DESCRIPTION
**Pull request by osala-eng**
## Description

Please include a summary of the change, e.g. what the new feature # is and/or what bug # it fixes. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes/Implements #(issue/feature)
Fixes page goes to login page on page reload even if you have already logged in.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor and code improvement (non-breaking change which improves code quality and/or performance)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [X] Tests
- [ ] Other chores such as maintenance

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration by modifying the list below.

- [ ] Test A
- [ ] Test B
- [ ] Test C

**Test Configuration**:

Please describe the test setup. List them below as bullet points.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [x] The commit message follows the convention of this project

### CODE REVIEW

File: source/docq/support/auth_utils.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+cached_sessions:TTLCache[str, bytes] = TTLCache(*CACHE_CONFIG)
+session_data:TTLCache[str, str]= TTLCache(*CACHE_CONFIG)
+
+def _set_cookie(cookie: str) -> None:
+    """Set client cookie for authentication."""
+    try:
+        expiry = datetime.now() + timedelta(hours=4)
+        html(f"""
+        <script>
+            document.cookie = "{COOKIE_NAME}={cookie}; expires={expiry.strftime('%a, %d %b %Y %H:%M:%S GMT')}; path=/;";
```
Comment (janaka) [inline source/docq/support/auth_utils.py:36]: Have you checked what `Domain` is set to? I think when not set explicitly the default behaviour is secure i.e. it should be auto-set to what ever domain the request came on. This makes sure an app on another domain cannot read these secret cookies.

**Review by janaka — CHANGES_REQUESTED** [review]
Nice I think this is looking pretty good for our needs now. I left a few comments.

@cwang this one can do with another pair of eyes. Replay attacks are the main exposure I was worried about leaving open but encrypting the hash mitigates that fine.

File: source/docq/manage_settings.py [hunk 2]
```diff
@@ -31,15 +31,16 @@ def _init() -> None:
...
 
 
-def _get_sqlite_file(user_id: int = None) -> str:
+def _get_sqlite_file(user_id: int | None = None) -> str:
     """Get the sqlite file for the given user."""
     return get_sqlite_usage_file(user_id) if user_id else get_sqlite_system_file()
 
 
-def _get_settings(user_id: int = None) -> dict:
+def _get_settings(user_id: int | None = None) -> dict:
```
[the remark this answers is not in the document]
Comment (osala-eng) [reply source/docq/manage_settings.py:39]: Yes, this gets called when calling `_get_sqlite_file()`

File: source/docq/support/auth_utils.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+from cachetools import TTLCache
+from cryptography.fernet import Fernet
+from streamlit.components.v1 import html
+from streamlit.web.server.websocket_headers import _get_websocket_headers
+
+from ..config import COOKIE_NAME, ENV_VAR_COOKIE_SECRET_KEY
+
+CACHE_CONFIG = (1024 * 1, 60 * 60 * 24)
+AUTH_KEY = Fernet.generate_key()
+COOKIE_SECRET = "secret_key"
```
[the remark this answers is not in the document]
Comment (osala-eng) [reply source/docq/support/auth_utils.py:21]: Okay, I'm renaming the secret key and adding a validation for the number of bits.

File: source/docq/support/auth_utils.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+with suppress(FileNotFoundError):
+    COOKIE_SECRET = st.secrets[ENV_VAR_COOKIE_SECRET_KEY]
+
+"""Session Cache"""
+cached_sessions:TTLCache[str, bytes] = TTLCache(*CACHE_CONFIG)
+session_data:TTLCache[str, str]= TTLCache(*CACHE_CONFIG)
```
[the remark this answers is not in the document]
Comment (osala-eng) [reply source/docq/support/auth_utils.py:28]: I'm adding a variable at the top to control both the local cache duration and the client side cookie expiry.

File: source/docq/support/auth_utils.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        if cookies is not None:
+            cookie = cookies.get(COOKIE_NAME)
+            if cookie is None:
+                return None
+            if cookie not in cached_sessions:
+                return None
+            if not _verify_hmac(session_data[cookie], cookie):
+                log.warning("Session ID not verified: %s", cookie)
+                return None
+            return cookie
```
[the remark this answers is not in the document]
Comment (osala-eng) [reply source/docq/support/auth_utils.py:105]: Yes, if a cookie is invalid a login screen is presented.

File: source/docq/support/auth_utils.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    """Verify credibility of HMAC hash."""
+    return hmac.compare_digest(
+        _create_hmac(msg),
+        digest
+    )
+
+def generate_session_id(length: int = 32) -> str:
+    """Generate a secure and unique session_id."""
+    id_ = token_hex(length // 2)
+    hmac_ = _create_hmac(id_)
```
[the remark this answers is not in the document]
Comment (osala-eng) [reply source/docq/support/auth_utils.py:84]: Apart from the `key` and `salt` the rest basically depends on our needs. i.e We could lock the sessions to a specific `IP address`, e.t.c

File: source/docq/support/auth_utils.py (hunk 1 above)
Comment (osala-eng) [reply source/docq/support/auth_utils.py:36]: Yes, this is locked to the domain that the request came from.

File: source/docq/support/auth_utils.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    COOKIE_SECRET = st.secrets[ENV_VAR_COOKIE_SECRET_KEY]
+
+"""Session Cache"""
+cached_sessions:TTLCache[str, bytes] = TTLCache(*CACHE_CONFIG)
+session_data:TTLCache[str, str]= TTLCache(*CACHE_CONFIG)
+
+def _set_cookie(cookie: str) -> None:
+    """Set client cookie for authentication."""
+    try:
+        expiry = datetime.now() + timedelta(hours=4)
```
[the remark this answers is not in the document]
Comment (osala-eng) [reply source/docq/support/auth_utils.py:33]: Not yet a sliding window, let me add that.

File: web/utils/handlers.py [hunk 8]
```diff
@@ -40,32 +41,49 @@
...
-            {
-                SessionKeyNameForSettings.SYSTEM.name: manage_settings.get_system_settings(),
-                SessionKeyNameForSettings.USER.name: manage_settings.get_user_settings(result[0]),
-            }
-        )
+                {
+                    SessionKeyNameForSettings.SYSTEM.name: manage_settings.get_system_settings(),
+                    SessionKeyNameForSettings.USER.name: manage_settings.get_user_settings(result[0]),
+                }
+            )
```
Comment (cwang) [inline web/utils/handlers.py:60]: Some bad indentations between L48 and L60?

File: source/docq/support/utils.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+"""Cache user sessions."""
```
Comment (cwang) [inline source/docq/support/utils.py:1]: Why cannot this be put in the `sessions.py` file in the same directory, given the same purpose?

File: source/docq/support/auth_utils.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
+"""Cache user sessions."""
```
Comment (cwang) [inline source/docq/support/auth_utils.py:1]: I'm surprised that you had to do it by using crypto apis directly - presumably none of the encrypted cookie plugins would do the job?

File: source/docq/manage_users.py [hunk 11]
```diff
@@ -13,7 +13,9 @@
 from . import manage_settings as msettings
 from .config import SpaceType
 from .domain import SpaceKey
+from .support.auth_utils import cache_auth
 from .support.store import get_sqlite_system_file
+from .support.utils import cache_auth
```
Comment (cwang) [inline source/docq/manage_users.py:18]: Which one is actually used, between L16 and L18?

**Review by cwang — COMMENTED** [review]
Left a few comments. More significantly, when running this branch, I'm getting

Screenshot 2023-09-06 at 15 31 42

(having done `poe install`)

File: source/docq/support/auth_utils.py (hunk 10 above)
Comment (janaka) [reply source/docq/support/auth_utils.py:1]: which plugins were you thinking specifically?

Comment (cwang) [reply source/docq/support/auth_utils.py:1]: Something like https://pypi.org/project/streamlit-cookies-manager/ but there're more

Comment (janaka) [reply source/docq/support/auth_utils.py:1]: Note: added some info here https://github.com/docqai/docq/issues/37#issuecomment-1712877001

**Comment by osala-eng** [comment]
Moving this to #97

**Pull request closed without merging**